cores/esp32/esp32-hal-uart.c

```diff
@@ -306,7 +303,7 @@ void uartFlushTxOnly(uart_t* uart, bool txOnly)
     }
     
     UART_MUTEX_LOCK();
-    ESP_ERROR_CHECK(uart_wait_tx_done(uart->num, portMAX_DELAY));
+    while(!uart_ll_is_tx_idle(UART_LL_GET_HW(uart->num)));
```

Comment:
Could somebody explain why exactly was the `uart_wait_tx_done()` call replaced here with the `uart_ll_is_tx_idle()` loop?

The ModbusMaster library is now broken in ESP32S2 v2.0.3, and it was working correctly in v2.0.2 (checked, I switch between the two versions at will). Our project, like many others, uses RS485 and flips the data-enable and read-enable pins, and uses flush() to wait until TX done before switching to RX mode. I still have to do more tests, but this change in particular seems the most likely culprit.

Reply:
@avillacis 

Have you read the issues related to this PR? They are related to MODBUS and `flush()` issues.

In summary:
`uart_wait_tx_done()` returns as soon as data is stored in TXRingBuffer, before the TX FIFO is really empty.

`uart_ll_is_tx_idle()` returns true just when TX FIFO is completely empty. 
The code for it is here - but it is the same for all other SoC, thus, the issue you see should happen for any ESP32-xx
https://github.com/espressif/esp-idf/blob/master/components/hal/esp32s2/include/hal/uart_ll.h#L714-L717

I tested `flush()` and I'm sure it works fine.
Maybe the issue you see isn't related to this PR.

Reply:
Objection withdrawn. With access to the actual hardware, I found out that our RS485 card was echoing back request commands, and this echo was previously hidden by the `uart_set_mode()` call. Now I have manually added the call in our own code, and the issue was fixed.

However, is there any upstream bug report documenting this issue with `uart_wait_tx_done()`? Is it a true software bug, or a documentation error?

Reply:
Apparently `uart_wait_tx_done()` works correctly when used with `uart_set_mode(uart_num, UART_MODE_RS485_HALF_DUPLEX)`, considering an IDF environment.
 
But for Arduino, it isn't useful because of the its generic UART usage and APIs.
Thus we had to change it for the LL API.

Reply:
By default, IDF starts UART with  `uart_set_mode(uart_num, enumerator UART_MODE_UART)` and in this case, it returns as soon as data is stored in TXRingBuffer, before the TX FIFO is really empty.